app/gui2/src/util/config.ts

```diff
@@ -0,0 +1,477 @@
+    let unrecognized: string[] = []
+    const addUnrecognized = (name: string) => {
+      unrecognized.push(stack.concat([name]).join('.'))
+    }
+    for (const [key, value] of Object.entries(config)) {
+      if (typeof value === 'string') {
+        const option = this.options[key]
+        if (option == null) {
+          addUnrecognized(key)
+        } else {
+          option.load(value)
+        }
+      } else {
+        const group = this.groups[key]
+        if (group == null) {
+          addUnrecognized(key)
+        } else {
+          const subStack = stack.concat([key])
+          unrecognized = unrecognized.concat(group.load(value, subStack))
+        }
+      }
+    }
+    return unrecognized
+  }
+
+  loadAllAndDisplayHelpIfUnsuccessful(configs: (StringConfig | null | undefined)[]): boolean {
+    let unrecognized: string[] = []
+    for (const config of configs) {
+      if (config != null) {
+        unrecognized = unrecognized.concat(this.load(config))
+      }
+    }
+    if (unrecognized.length !== 0) {
+      console.error(`Unrecognized configuration parameters: ${unrecognized.join(', ')}.`)
+      this.showConfigOptions(unrecognized)
+      return false
+    } else {
+      return true
+    }
+  }
+
+  stringify(): StringConfig {
+    const config: StringConfig = {}
+    const groupsEntries = Object.entries(this.groups)
+    if (groupsEntries.length > 0) {
+      config.groups = {}
+      for (const [name, group] of groupsEntries) {
+        config.groups[name] = group.stringify()
+      }
+    }
+    const optionsEntries = Object.entries(this.options)
+    if (optionsEntries.length > 0) {
+      config.options = {}
+      for (const [name, option] of optionsEntries) {
+        config.options[name] = option.value.toString()
+      }
+    }
+    return config
+  }
+
+  prettyPrint(indent = 0): string {
+    // The number is used for sorting in ordering to put options before groups.
+    const entries: [string, number, string][] = []
+    const optionsEntries = Object.entries(this.options)
+    if (optionsEntries.length > 0) {
+      for (const [name, option] of optionsEntries) {
+        entries.push([name, 0, option.value.toString()])
+      }
+    }
+    const groupsEntries = Object.entries(this.groups)
+    if (groupsEntries.length > 0) {
+      for (const [name, group] of groupsEntries) {
+        entries.push([name, 1, '\n' + group.prettyPrint(indent + 1)])
+      }
+    }
+    entries.sort()
+    return entries.map(([name, , value]) => ' '.repeat(2 * indent) + name + ': ' + value).join('\n')
+  }
+
+  optionsRecursive(): AnyOption[] {
+    const options: AnyOption[] = []
+    for (const option of Object.values(this.options)) {
+      options.push(option)
+    }
+    for (const group of Object.values(this.groups)) {
+      options.push(...group.optionsRecursive())
+    }
+    return options
+  }
+
+  /** Show an error dialog with a list of options. */
+  showConfigOptions(unknownOptions?: string[]) {
```

Comment:
Again mixing data (config object) and view (help page) into one.

Reply:
actually, the only time it is called is in `loadAllAndDisplayHelpIfUnsuccessful`